src/main/java/org/opentripplanner/raptor/rangeraptor/TimeBasedRoutingStrategy.java

```diff
@@ -0,0 +1,249 @@
+package org.opentripplanner.raptor.rangeraptor;
+
+import java.util.function.IntConsumer;
+import javax.annotation.Nonnull;
+import org.opentripplanner.raptor.rangeraptor.internalapi.RoundProvider;
+import org.opentripplanner.raptor.rangeraptor.internalapi.RoutingStrategy;
+import org.opentripplanner.raptor.rangeraptor.internalapi.SlackProvider;
+import org.opentripplanner.raptor.rangeraptor.internalapi.WorkerLifeCycle;
+import org.opentripplanner.raptor.rangeraptor.transit.TransitCalculator;
+import org.opentripplanner.raptor.spi.RaptorAccessEgress;
+import org.opentripplanner.raptor.spi.RaptorConstrainedTripScheduleBoardingSearch;
+import org.opentripplanner.raptor.spi.RaptorTimeTable;
+import org.opentripplanner.raptor.spi.RaptorTripSchedule;
+import org.opentripplanner.raptor.spi.RaptorTripScheduleBoardOrAlightEvent;
+import org.opentripplanner.raptor.spi.RaptorTripScheduleSearch;
+import org.opentripplanner.raptor.spi.TransitArrival;
+import org.opentripplanner.routing.algorithm.raptoradapter.transit.request.TripScheduleBoardSearch;
+
+/**
+ * This class contains code which is shared by all time-dependent routing strategies. It also
+ * defines abstract methods, which the individual strategies must implement.
+ */
+public abstract class TimeBasedRoutingStrategy<T extends RaptorTripSchedule>
+  implements RoutingStrategy<T> {
+
+  private final SlackProvider slackProvider;
+  private final TransitCalculator<T> calculator;
+  private final RoundProvider roundProvider;
+  private boolean inFirstIteration = true;
+  private boolean hasTimeDependentAccess = false;
+  protected RaptorTimeTable<T> timeTable;
+  protected RaptorTripScheduleSearch<T> tripSearch;
+
+  protected TimeBasedRoutingStrategy(
+    SlackProvider slackProvider,
+    TransitCalculator<T> calculator,
+    RoundProvider roundProvider,
+    WorkerLifeCycle subscriptions
+  ) {
+    this.slackProvider = slackProvider;
+    this.calculator = calculator;
+    this.roundProvider = roundProvider;
+
+    subscriptions.onIterationComplete(() -> inFirstIteration = false);
+  }
+
+  @Override
+  public final void prepareForTransitWith(RaptorTimeTable<T> timeTable) {
+    this.timeTable = timeTable;
+    this.tripSearch = createTripSearch(timeTable);
+    prepareForTransitWith();
+  }
+
+  protected abstract void prepareForTransitWith();
+
+  /**
+   * Get the current boarding previous transit arrival. This is used to look up any guaranteed
+   * transfers.
+   */
+  protected abstract TransitArrival<T> previousTransit(int boardStopIndex);
+
+  @Override
+  public final void board(
+    int stopIndex,
+    int stopPos,
+    int boardSlack,
+    boolean hasConstrainedTransfer,
+    RaptorConstrainedTripScheduleBoardingSearch<T> txSearch
+  ) {
+    // MC Raptor have many, while RR have one boarding
+    forEachBoarding(
```

Comment:
Should we inline this to the different strategies, as we have only one boarding for single criteria raptor?